### PR TITLE
fix(psync): log the error if payment method retrieve fails in the `psync flow`

### PR DESCRIPTION
**Pull request by ShankarSinghC**
## Type of Change

- [x] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

-> When a payment is created with save card we store the card in the locker and also a payment method entry is created in the payment method table. And the corresponding payment_method_id will be stored in the payment_attempt table.
-> And later when that payment method is deleted, psycn is failing as we are trying to fetch the payment method data from the payment_method table in the psync flow.
-> The error occurred because of the recent change that was trying to fetch the payment method in the psync flow using the payment_method_id in the payment attempt.

This is fixed by adding a validation for the error that occurs during fetching the payment method data from the db. If it is not found in db error then just log the error (do not propagate it), any other error needs to be returned as it is.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Create a payment with save card
image
```
{
    "amount": 100,
    "amount_to_capture": 100,
    "currency": "USD",
    "confirm": true,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "customer_id": "aaa",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "setup_future_usage": "on_session",
    "return_url": "http://127.0.0.1:4040",
    "customer_acceptance": {
        "acceptance_type": "offline",
        "accepted_at": "1963-05-03T04:07:52.723Z",
        "online": {
            "ip_address": "in sit",
            "user_agent": "amet irure esse"
        }
    },
    "payment_method": "card",
    "payment_method_data": {
        "card": {
            "card_number": "4242424242424242",
            "card_exp_month": "03",
            "card_exp_year": "2030",
            "card_holder_name": "joseph Doe",
            "card_cvc": "838"
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "PiX",
            "last_name": "ss"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "PiX"
        }
    },
    "browser_info": {
        "user_agent": "Mozilla\/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/70.0.3538.110 Safari\/537.36",
        "accept_header": "text\/html,application\/xhtml+xml,application\/xml;q=0.9,image\/webp,image\/apng,*\/*;q=0.8",
        "language": "nl-NL",
        "color_depth": 24,
        "screen_height": 723,
        "screen_width": 1536,
        "time_zone": 0,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "125.0.0.1"
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    }
}
```
Delete the saved payment mehtod
```
curl --location --request DELETE 'http://localhost:8080/payment_methods/{{pm_id}}' \
--header 'Accept: application/json' \
--header 'api-key: abc'
```
payment method entry being deleted from db
image
Retrieve the above payment
```
curl --location 'http://localhost:8080/payments/pay_7VFY24kczcTpt0xQfU8M?force_sync=true' \
--header 'Accept: application/json' \
--header 'api-key: dev_uU4PZwC903VYtQnTBvTxgU0QFDmah0mcPN5qehIq7tS0oapEfBBVp96oaIHyjDdM'
```
image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/core/payments/operations/payment_status.rs [hunk 1]
```diff
@@ -399,14 +399,20 @@ async fn get_tracker_for_sync<
...
-                    .await
-                    .to_not_found_response(errors::ApiErrorResponse::PaymentMethodNotFound)
-                    .attach_printable("error retrieving payment method from DB")?,
-            )
+            match db.find_payment_method(payment_method_id).await {
+                Ok(payment_method) => Some(payment_method),
+                Err(error) => {
+                    if error.current_context().is_db_not_found() {
+                        logger::info!("Error retrieving payment method from DB {:?}", error);
+                        None
```
Comment (Chethan-rao) [inline crates/router/src/core/payments/operations/payment_status.rs:410]: Is this error log required? If required then change it to something like
`Payment Method was not found in db`. It's not an error from db right?

Comment (ShankarSinghC) [reply crates/router/src/core/payments/operations/payment_status.rs:410]: Yes, it is required log. Sure I will change the error message.

File: crates/router/src/core/payments/operations/payment_status.rs [hunk 2]
```diff
@@ -399,14 +399,20 @@ async fn get_tracker_for_sync<
...
+                    } else {
+                        Err(error).change_context(errors::ApiErrorResponse::InternalServerError)?
+                    }
```
Comment (Chethan-rao) [inline crates/router/src/core/payments/operations/payment_status.rs:413]: Might as well add an attach printable here saying pm retrieval from db failed?

**Review by Chethan-rao — APPROVED** [review]

File: crates/router/src/core/payments/operations/payment_status.rs [hunk 3]
```diff
@@ -399,14 +399,22 @@ async fn get_tracker_for_sync<
...
-                    .await
-                    .to_not_found_response(errors::ApiErrorResponse::PaymentMethodNotFound)
-                    .attach_printable("error retrieving payment method from DB")?,
-            )
+            match db.find_payment_method(payment_method_id).await {
+                Ok(payment_method) => Some(payment_method),
+                Err(error) => {
+                    if error.current_context().is_db_not_found() {
+                        logger::info!("Payment Method not found in db {:?}", error);
+                        None
```
[the remark this answers is not in the document]
Comment (ShankarSinghC) [reply crates/router/src/core/payments/operations/payment_status.rs:410]: Yeah, I will create an issue for the soft delete implementation. We will pick this up.
This pr is a temporary fix for the merchants testing on sandbox.

[the remark this answers is not in the document]
Comment (Narayanbhat166) [reply crates/router/src/core/payments/operations/payment_status.rs:410]: Please link the issue here once you have created it

[the remark this answers is not in the document]
Comment (ShankarSinghC) [reply crates/router/src/core/payments/operations/payment_status.rs:410]: Sure

[the remark this answers is not in the document]
Comment (ShankarSinghC) [reply crates/router/src/core/payments/operations/payment_status.rs:410]: Here is the issue for the above https://github.com/juspay/hyperswitch/issues/4324

**Review by Narayanbhat166 — APPROVED** [review]

**Pull request merged**